Crash when snarfing empty selection using mouse chord

To reproduce, cut and paste using mouse chord, without moving the mouse cursor. It doesn't crash if the cursor is at the beginning of the file.
```
panic: acme: textsetselect p0=315 p1=0 q0=315 q1=0 t.org=0 nchars=906

goroutine 38 [running, locked to thread]:
main.(*Text).SetSelect(0xc000194518, 0x13b, 0x0)
	/home/fhs/go/src/github.com/rjkroege/edwood/text.go:1288 +0x396
main.(*Text).Select(0xc000194518)
	/home/fhs/go/src/github.com/rjkroege/edwood/text.go:1176 +0x49e
main.MovedMouse(0x2ce, 0x10a, 0x1, 0x3f20b8ea)
	/home/fhs/go/src/github.com/rjkroege/edwood/acme.go:422 +0x48b
main.mousethread(0x69dde0, 0xc00009c118)
	/home/fhs/go/src/github.com/rjkroege/edwood/acme.go:301 +0x304
created by main.main.func1
	/home/fhs/go/src/github.com/rjkroege/edwood/acme.go:173 +0x82b
```

--- a/file.go
+++ b/file.go
@@ -527,13 +527,17 @@ func (f *File) Seq() int {
 	return f.seq
 }
 
+// Undo undoes edits if isundo is true or redoes edits if isundo is false.
+// It returns the new selection q0, q1 and a bool indicating if the
+// returned selection is meaningful.
+//
 // TODO(rjk): Separate Undo and Redo for better alignment with undo.Buffer
 // TODO(rjk): This Undo implementation may Undo/Redo multiple changes.
 // The number actually processed is controlled by mutations to File.seq.
 // This does not align with the semantics of undo.Buffer.
 // Each "Mark" needs to have a seq value provided.
 // TODO(rjk): Consider providing the target seq value as an argument.
-func (f *File) Undo(isundo bool) (q0p, q1p int) {
+func (f *File) Undo(isundo bool) (q0, q1 int, ok bool) {
 	var (
 		stop           int
 		delta, epsilon *[]*Undo
@@ -577,8 +581,9 @@ func (f *File) Undo(isundo bool) (q0p, q1p int) {
 			for _, text := range f.text {
 				text.deleted(u.p0, u.p0+u.n)
 			}
-			q0p = u.p0
-			q1p = u.p0
+			q0 = u.p0
+			q1 = u.p0
+			ok = true
 		case Insert:
 			f.seq = u.seq
 			f.Uninsert(epsilon, u.p0, u.n)
@@ -588,8 +593,9 @@ func (f *File) Undo(isundo bool) (q0p, q1p int) {
 			for _, text := range f.text {
 				text.inserted(u.p0, u.buf)
 			}
-			q0p = u.p0
-			q1p = u.p0 + u.n
+			q0 = u.p0
+			q1 = u.p0 + u.n
+			ok = true
 		case Filename:
 			// TODO(rjk): If I have a zerox, does undo a filename change update?
 			f.seq = u.seq
@@ -605,7 +611,7 @@ func (f *File) Undo(isundo bool) (q0p, q1p int) {
 	if isundo {
 		f.seq = 0
 	}
-	return q0p, q1p
+	return q0, q1, ok
 }
 
 // Reset removes all Undo records for this File.

--- a/wind.go
+++ b/wind.go
@@ -388,7 +388,9 @@ func (w *Window) Delete() {
 func (w *Window) Undo(isundo bool) {
 	w.utflastqid = -1
 	body := &w.body
-	body.q0, body.q1 = body.file.Undo(isundo)
+	if q0, q1, ok := body.file.Undo(isundo); ok {
+		body.q0, body.q1 = q0, q1
+	}
 
 	// TODO(rjk): Is this absolutely essential.
 	body.Show(body.q0, body.q1, true)

--- a/wind_test.go
+++ b/wind_test.go
@@ -1,0 +1,48 @@
+package main
+
+import "testing"
+
+// TestWindowUndoSelection checks text selection change after undo/redo.
+// It tests that selection doesn't change when undoing/redoing
+// using nil delta/epsilon, which fixes https://github.com/rjkroege/edwood/issues/230.
+func TestWindowUndoSelection(t *testing.T) {
+	var (
+		word = Buffer("hello")
+		p0   = 3
+		undo = &Undo{
+			t:   Insert,
+			buf: word,
+			p0:  p0,
+			n:   word.nc(),
+		}
+	)
+	for _, tc := range []struct {
+		name           string
+		isundo         bool
+		q0, q1         int
+		wantQ0, wantQ1 int
+		delta, epsilon []*Undo
+	}{
+		{"undo", true, 14, 17, p0, p0 + word.nc(), []*Undo{undo}, nil},
+		{"redo", false, 14, 17, p0, p0 + word.nc(), nil, []*Undo{undo}},
+		{"undo (nil delta)", true, 14, 17, 14, 17, nil, nil},
+		{"redo (nil epsilon)", false, 14, 17, 14, 17, nil, nil},
+	} {
+		w := &Window{
+			body: Text{
+				q0: tc.q0,
+				q1: tc.q1,
+				file: &File{
+					b:       Buffer("This is an example sentence.\n"),
+					delta:   tc.delta,
+					epsilon: tc.epsilon,
+				},
+			},
+		}
+		w.Undo(tc.isundo)
+		if w.body.q0 != tc.wantQ0 || w.body.q1 != tc.wantQ1 {
+			t.Errorf("%v changed q0, q1 to %v, %v; want %v, %v",
+				tc.name, w.body.q0, w.body.q1, tc.wantQ0, tc.wantQ1)
+		}
+	}
+}

